src/async_operation.c

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <stddef.h>
+#include <string.h>
+#include "azure_c_shared_utility/optimize_size.h"
+#include "azure_c_shared_utility/gballoc.h"
+#include "azure_c_shared_utility/xlogging.h"
+#include "azure_uamqp_c/async_operation.h"
+
+typedef struct ASYNC_OPERATION_INSTANCE_TAG
+{
+    ASYNC_OPERATION_CANCEL_HANDLER_FUNC async_operation_cancel_handler;
+} ASYNC_OPERATION_INSTANCE;
+
+ASYNC_OPERATION_HANDLE async_operation_create(ASYNC_OPERATION_CANCEL_HANDLER_FUNC async_operation_cancel_handler, size_t context_size)
+{
+    ASYNC_OPERATION_INSTANCE* async_operation;
+
+    if (async_operation_cancel_handler == NULL)
+    {
+        /* Codes_SRS_ASYNC_OPERATION_01_002: [ If `async_operation_cancel_handler` is NULL, `async_operation_create` shall fail and return NULL.]*/
+        LogError("Cannot allocate memory for async operation");
+    }
+    else if (context_size < sizeof(ASYNC_OPERATION_INSTANCE))
+    {
+        /* Codes_SRS_ASYNC_OPERATION_01_003: [ If `context_size` is less than the size of the `async_operation_cancel_handler` argument, `async_operation_create` shall fail and return NULL.]*/
+        LogError("Context size too small");
+    }
+    else
+    {
+        async_operation = (ASYNC_OPERATION_INSTANCE*)malloc(context_size);
+        if (async_operation == NULL)
+        {
+            /* Codes_SRS_ASYNC_OPERATION_01_004: [ If allocating memory for the new asynchronous operation instance fails, `async_operation_create` shall fail and return NULL.]*/
+            LogError("Cannot allocate memory for async operation");
+        }
+        else
+        {
+            /* Codes_SRS_ASYNC_OPERATION_01_001: [ `async_operation_create` shall return a non-NULL handle to a newly created asynchronous operation instance.]*/
+            async_operation->async_operation_cancel_handler = async_operation_cancel_handler;
+            goto all_ok;
```

Comment:
>goto all_ok; [](start = 12, length = 12)

I don't think this goto is providing much value.  without it you just have to have async_operation = NULL in two more places.  There's not any deallocation or save by it.

Reply:
It can go either way. I removed it anyhow this time :-).

---
In reply to: [142210232](https://github.com/Azure/azure-uamqp-c/pull/185#discussion_r142210232) [](ancestors = 142210232)